foyer-storage/src/generic.rs

```diff
@@ -353,77 +353,69 @@ where
-
-                let res = match read_entry::<K, V>(buf.as_ref()) {
-                    Ok((key, value)) => {
-                        self.inner.metrics.op_bytes_get.inc_by(buf.len() as u64);
-                        Ok(Some(CachedEntry::Owned {
-                            key: Box::new(key),
-                            value: Box::new(value),
-                        }))
-                    }
-                    Err(e) => {
-                        // Remove index if the storage layer fails to get it (because of entry magic mismatch).
-                        self.inner.catalog.remove(key);
-                        Err(e)
-                    }
-                };
-
-                self.inner
-                    .metrics
-                    .op_duration_get_hit
-                    .observe(now.elapsed().as_secs_f64());
+        async move {
+            let (_sequence, index) = match index_value {
+                Some(item) => item.consume(),
+                None => {
+                    inner.metrics.op_duration_get_miss.observe(now.elapsed().as_secs_f64());
+                    return Ok(None);
+                }
+            };
 
-                res
+            match index {
+                crate::catalog::Index::Inflight { key, value } => {
+                    let value = value.clone();
+
+                    inner.metrics.op_duration_get_hit.observe(now.elapsed().as_secs_f64());
+
+                    Ok(Some(CachedEntry::Shared { key, value }))
+                }
+                crate::catalog::Index::Region { view } => {
+                    let region = view.id();
+
+                    inner.region_manager.record_access(region);
+                    let region = inner.region_manager.region(region);
+
+                    // TODO(MrCroxx): read value only
+                    let buf = match region.load(view).await? {
+                        Some(buf) => buf,
+                        None => {
+                            // Remove index if the storage layer fails to get it (because of region version mismatch).
+                            // inner.catalog.remove(index.key());
+                            inner.metrics.op_duration_get_miss.observe(now.elapsed().as_secs_f64());
+                            return Ok(None);
+                        }
+                    };
+
+                    let res = match read_entry::<K, V>(buf.as_ref()) {
+                        Ok((key, value)) => {
+                            inner.metrics.op_bytes_get.inc_by(buf.len() as u64);
+                            Ok(Some(CachedEntry::Owned {
+                                key: Box::new(key),
+                                value: Box::new(value),
+                            }))
+                        }
+                        Err(e) => {
+                            // Remove index if the storage layer fails to get it (because of entry magic mismatch).
+                            // inner.catalog.remove(key);
```

Comment:
The only place that is non-trival. Can we remove the reference of key to make this future static?